### PR TITLE
react: function type as a state in useState

### PR DESCRIPTION
**Pull request by zepatrik**
The issue is that you have to provide a `lazyInitialState` to have a state of a function type in order to work.

``` ts
// this is accepted by current type definition but react will think it is a lazy state
const [callback, setCallback] = useState<() => void>(() => undefined)

// this should be forced by the type definition
const [callback, setCallback] = useState<() => void>(() => () => undefined)
```

I'm having a hard time typing this correctly as for some reason typescript narrows down the types when relying on inference with a conditional type. Also tried to use overloading which did not work either. Not sure if that is a bug in typescript or if I am just to stupid.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://reactjs.org/docs/hooks-reference.html#lazy-initial-state
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by ferdaber** [comment]
React doesn’t support bare functions as state, if you do want that you should wrap it inside an object so that it’s a method of the object.

**Comment by ferdaber** [comment]
The required typing is a bit complex, here's a past PR that broke some things for past reference: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36080

**Comment by zepatrik** [comment]
Right, thanks for the reference.
So what is the desired way to go? Either the typing should disallow function as state (technically not right) or force you to use lazy initial state and dispatch form of setState.

**Comment by ferdaber** [comment]
If you can make the tests work with your changes, I'd say go for it. It may be a rare use case but we should still support it since the runtime allows it.

File: types/react/index.d.ts [hunk 1]
```diff
@@ -833,7 +833,8 @@ declare namespace React {
      * @version 16.8.0
      * @see https://reactjs.org/docs/hooks-reference.html#usestate
      */
-    function useState<S>(initialState: S | (() => S)): [S, Dispatch<SetStateAction<S>>];
+    // function useState<S extends Function>(lazyState: () => S): [S, Dispatch<SetStateAction<S>>];
+    function useState<S>(initialState: S extends Function ? () => S : (() => S) | S): [S, Dispatch<SetStateAction<S>>];
```
Comment (Jessidhia) [inline types/react/index.d.ts]: `Function` is not allowed by the linter; and it seems that using a conditional type here is causing the compiler to infer stricter / const-like types for the input 😞

I haven't tested it, but try this:

```ts
function useState<T>(initializer:
  | () => T
  | Exclude<T, (...args: any[]) => any>
): [T, Dispatch<SetStateAction<T>>]
```

You also need to do the same change on `SetStateAction`.

Comment (zepatrik) [reply types/react/index.d.ts]: Right, it would work if the compiler wouldn't infer stricter. I could not find any resources on why this is happening.

**Comment by zepatrik** [comment]
So closing this as I am not able to do this right. Inference with conditional types in typescript behaves quite weird.

**Pull request closed without merging**